### PR TITLE
[move-prover] Using immutable collections for SetDomain and MapDomain

### PR DESCRIPTION
**Pull request by wrwg**
This reimplements the collection types for DFA using base types from the [`im` crate](https://docs.rs/im/15.0.0/im/).

In DFA, we often have types like maps or sets which are copied and refined with each basic block. Traditional Rust collections are not well suited for this, because for each instruction we need to make a deep clone of the input domains. The `im` types are designed with structure sharing in mind. Clone is O(1), and subsequent write leads only to a delta with a reference to the original value.

The API of those collections is slightly different which lead to some adoptions in our code. The PR also attempts to complete the wrapping traits for SetDomain and MapDomain such that they (mostly) behave as the underlying representation.

## Motivation

Make the DFA framework more sensible.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Existing tests.

This PR also fixes a test failure which exists in main and breaks the new `disable_inv.move` test for invariants v2. This test needs it's separate baseline, since the other features produce different errors.

## Related PRs

NA

### CODE REVIEW

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A im 15.0.0 (direct third-party, crates.io)
    * features: [none]
  A bitmaps 2.1.0 (transitive third-party, crates.io)
    * features: default, std
  A rand_xoshiro 0.4.0 (transitive third-party, crates.io)
    * features: [none]
  A sized-chunks 0.6.5 (transitive third-party, crates.io)
    * features: default, std
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A im 15.0.0 (direct third-party, crates.io)
    * features: [none]
  A bitmaps 2.1.0 (transitive third-party, crates.io)
    * features: default, std
  A rand_xoshiro 0.4.0 (transitive third-party, crates.io)
    * features: [none]
  A sized-chunks 0.6.5 (transitive third-party, crates.io)
    * features: default, std
```

File: language/move-prover/bytecode/src/dataflow_domains.rs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        (ordmap! {k => v}).into()
+    }
+
+    /// Join `v` with self[k] if `k` is bound, insert `v` otherwise
+    pub fn insert_join(&mut self, k: K, v: V) -> JoinResult {
+        let mut change = JoinResult::Unchanged;
+        self.0
+            .entry(k)
+            .and_modify(|old_v| {
+                change = change.combine(old_v.join(&v));
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/dataflow_domains.rs:253]: I think

```suggestion
                change = old_v.join(&v);
```

is equivalent/slightly simpler.

**Review by sblackshear — APPROVED** [review]
Thanks, this looks great!!

File: language/move-prover/bytecode/src/dataflow_domains.rs (hunk 1 above)
Comment (wrwg) [reply language/move-prover/bytecode/src/dataflow_domains.rs:253]: Done

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A im 15.0.0 (direct third-party, crates.io)
    * features: [none]
  A bitmaps 2.1.0 (transitive third-party, crates.io)
    * features: default, std
  A rand_xoshiro 0.4.0 (transitive third-party, crates.io)
    * features: [none]
  A sized-chunks 0.6.5 (transitive third-party, crates.io)
    * features: default, std
```

**Pull request merged**